src/update.rs

```diff
@@ -0,0 +1,85 @@
+use std::env::consts;
+
+#[cfg(any(target_os = "windows", target_os = "linux"))]
+use regex::Regex;
+#[cfg(any(target_os = "windows", target_os = "linux"))]
+use self_update::backends::github::{ReleaseList, UpdateBuilder};
+use thiserror::Error;
+
+#[cfg(not(any(target_os = "windows", target_os = "linux")))]
+#[derive(Debug, Clone, Copy, Error)]
+pub enum UpdateError {
+    #[error("Not supported on the platform: {0}")]
+    NotSupported(&'static str),
+}
+
+#[cfg(any(target_os = "windows", target_os = "linux"))]
+#[derive(Debug, Error)]
+pub enum UpdateError {
+    #[error("Version tag not in the format of '(v)X.Y.Z'")]
+    VersionTag,
+    #[error("No release is available for the target")]
+    MissingRelease,
+    #[error("{0}")]
+    SelfUpdate(#[from] self_update::errors::Error),
+}
+
+#[allow(unused_variables)]
+#[allow(clippy::missing_panics_doc)]
+pub fn update(
+    bin_name: &str,
+    current_version: &str,
+    target_version: Option<&str>,
+) -> Result<(), UpdateError> {
+    #[cfg(not(any(target_os = "windows", target_os = "linux")))]
+    {
+        Err(UpdateError::NotSupported(consts::OS))
+    }
+    #[cfg(any(target_os = "windows", target_os = "linux"))]
+    {
+        let mut version = if let Some(version) = target_version {
+            let version = version.to_owned();
+            let re = Regex::new(r"^v?[0-9]+\.[0-9]+\.[0-9]+$").unwrap();
+            if !re.is_match(&version) {
+                return Err(UpdateError::VersionTag);
+            }
+            version
+        } else {
+            let releases = ReleaseList::configure()
+                .repo_owner("amtep")
+                .repo_name("ck3-tiger")
+                .with_target(consts::OS)
+                .build()?
+                .fetch()?;
+
+            releases.first().ok_or(UpdateError::MissingRelease)?.version.clone()
```

Comment:
I think this needs a few more smarts, to skip releases that are "draft" or "pre-release"

Reply:
Right now this is a little cursed by the limitation of `self_update` we discussed before. We can filter the releases and get the first desirable release, and specify that in the following updater. However, in doing so, it would be a forced-update even if it is not needed.

Reply:
I've looked deeper into `self_update` and it seems that draft and prerelease booleans are currently not used. This again requires upstream changes to function properly.